### PR TITLE
Add featured transforms in Link Control for Navigation Link items.

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Partially addresses #34041

Adds three transform buttons inside the link control for Navigation Links. The blocks to transform to are hardcoded; they're a selection of the options that might be most useful when expanding the Navigation to include more than just links: Search, Site Logo and Social Icons.

Adds a `renderControlBottom` render prop to LinkControl that allows adding extra controls to the bottom of the link popover.

## How has this been tested?

In the Navigation block, add a new link. Check that block transforms are showing in the link popover. 
Open the link popover for an existing navigation link that has a URL set. Transforms should not display.
Link popover should be unchanged for regular links in paragraphs etc..

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
Took it for a quick spin, and seeing these nice waiting-to-be-styled buttons in the transform dialog:
transforms

Honestly with some nice styles here, I imagine this could be pretty decent! If you feel happy about the code here, and @getdave is okay with it as well, I'd be happy to push some CSS tweaks to shine this one up! (Though I'd appreciate help adding the right icons and labels to the buttons!)

File: packages/block-editor/src/components/link-control/index.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/index.js]: My concern is, whilst this is _expedient_, it does lock us into a particular location for `children`. 

I wonder whether instead we have dedicated locations for rendering additional UI inside LinkControl. 

So in this case we want to add content to the bottom of the control so a property for this would be called `renderControlBottom` (or something). We could treat as a render prop and that would allow you to optionally pass content to be rendered there. Treating as a render prop would also give you the ability to pass some of the internal state of link control to the inserted component.

If later on we decide we need to insert content into another location we can add `renderControlTop` or whatever.

Comment (tellthemachines) [reply packages/block-editor/src/components/link-control/index.js]: Changed this to use render prop! I hate naming things so conveniently used your suggestion 😄

**Comment by tellthemachines** [comment]
Updated to add the block icon and proper title, as well as use the Button component which gives us at least a reset of user agent button styles. If we want these buttons to look like the inserter ones, I think we only need to move the icon on top of the text and add a bit of padding.

I'm thinking we also need a title for this section. Shall we go with "Or transform to block" as @jasmussen's design shows?

A few questions to consider:

Currently the three items on display are hardcoded. Will we ever want to have different blocks appearing? 

I asked this already on the issue, but adding it here again for visibility: do we want these transforms to always display in the link control for nav links, or do we want to hide them when different block types are already available from the quick inserter?

Also, do we want to hide these when a link is already set in the block? E.g.
Screen Shot 2021-10-25 at 5 35 00 pm

**Comment by jasmussen** [comment]
This is looking good!

Screenshot 2021-10-25 at 08 52 03

I'd like to add a separator, tweak the padding a little bit, and tinker with whether stacked buttons are better or not, like this:

Screenshot 2021-10-25 at 08 53 06

Mind if I push to the branch?

Speaking of, should we change the button defaults to logo, title, search? Depends a bit on whether we can land #35688 soon, if we can then Search might be a better experience than Social Links 🤔 

> Also, do we want to hide these when a link is already set in the block?

Screenshot 2021-10-25 at 08 52 13

My instinct would be to hide these, yes.

**Comment by jasmussen** [comment]
Actually I held off on pushing my changes because it felt like I was adding a tone of bespoke CSS in arbitrary places. So I'll hold off for now, and let me know if/how/where this would be appropriate. I managed to get it to look like this:
Screenshot 2021-10-25 at 09 14 29

That moves the "Open in new tab" to a toggle you can flip after the fact, not at the initial step. 

It assumes this markup snippet:

```
								renderControlBottom={ () => (
									<div className="link-control-transform">
										<h3 className="link-control-transform__subheading">
											{ __( 'Transform' ) }
										</h3>
										<div className="link-control-transform__items">
											<LinkControlTransforms
												block={ thisBlock }
												transforms={
													featuredTransforms
												}
											/>
										</div>
									</div>
								) }
```

I then added these pieces of CSS. Buttons:

```
.components-button.link-control-transform__item {
	flex-direction: column;
	gap: $grid-unit-10;
	height: auto;
	flex-basis: 33%;
}
```

Button container — I guess we might have a counterpart among the new components?

```
.link-control-transform__items {
	display: flex;
	justify-content: space-between;
}
```

Subheading, basically the same as [this](https://github.com/WordPress/gutenberg/blob/trunk/packages/interface/src/components/complementary-area/style.scss#L45) one, so it should probably be reused:

```
.link-control-transform__subheading {
    font-size: 11px;
    text-transform: uppercase;
    font-weight: 500;
    color: $gray-900;
    margin-bottom: 1.5em;
}
```

General container:

```
.link-control-transform {
	border-top: $border-width solid $gray-400;
	padding: 0 $grid-unit-20 $grid-unit10 $grid-unit-20;
}
```

As we look to the new components to absorb much of this bespoke CSS, it just feels wrong to add it. At the same time, it might be the most direct path to a solution. What do you think?

**Comment by tellthemachines** [comment]
Thanks for sharing those styles @jasmussen ! I incorporated them in the lastest changeset and it should be looking more or less as expected now.

> Subheading, basically the same as [this](https://github.com/WordPress/gutenberg/blob/trunk/packages/interface/src/components/complementary-area/style.scss#L45) one, so it should probably be reused:

We're already using `font-size: 11px; text-transform: uppercase; font-weight: 500;` in a number of different places, though the color is usually `$gray-700`. Perhaps at some point we should extract these styles to somewhere common, but right now we only have a bunch of instances in super specific places, none of which really makes sense to reuse here.

> That moves the "Open in new tab" to a toggle you can flip after the fact, not at the initial step.

There's no ready-made way to do this. "Open in new tab" is a default setting in LinkControl, and currently there's no way of setting it conditionally in that component. Ideally we'd be able to pass a prop to LinkControl to configure when to render "Open in new tab", but as that involves changing the LinkControl API, we should do it as a separate task. Cc. @getdave

File: packages/block-library/src/navigation-link/edit.js
Comment (talldan) [inline packages/block-library/src/navigation-link/edit.js]: I'm wondering if it should use a selector like `getBlockTransformItems`.

File: packages/block-library/src/navigation-link/edit.js
Comment (talldan) [inline packages/block-library/src/navigation-link/edit.js]: Could this one be un-nested? Or alternatively wrapped in `useCallback`?

In the past, I've seen a component defined within another component in this way cause unexpected behaviors in React. The reconciler can treat it as a new component on every render because the function reference changes each time. 

I actually haven't noticed any issues in testing, but I still think it's a pattern worth avoiding.

File: packages/block-library/src/navigation-link/edit.js
Comment (talldan) [inline packages/block-library/src/navigation-link/edit.js]: Not sure what the right class name would be. It should maybe use the `wp-block-navigation-link` class if its defined with the component. It could get quite wordy.

File: packages/block-library/src/navigation-link/edit.js
Comment (talldan) [inline packages/block-library/src/navigation-link/edit.js]: Maybe not needed for a first draft, but if these hard-coded blocks are unregistered by a site maintainer or plugin, it could be worth falling back to other available transforms.

**Review by talldan — COMMENTED** [review]
Works really well. A few little adjustments and it should be good to ship. I noticed in testing the styling doesn't seem quite right:
Screenshot 2021-10-26 at 4 01 21 pm

Maybe a missing flex direction.

**Comment by jasmussen** [comment]
Incredible work, and it's coming together so fast, this looks very close to me:
Screenshot 2021-10-26 at 11 23 26

I also appreciate that you used the flex components to make this happen, rather than the bespoke CSS I suggested! 
The only other piece I'd love for us to find a way to fix is making each item the same size:
Screenshot 2021-10-26 at 11 23 46

I solved that in the pseudo CSS by adding `flex-basis: 33%;`. But it appears the individual link transform items are meant to be flex, but are actually block:
Screenshot 2021-10-26 at 11 30 57

It appears partially intentional from the component side of things, there's even a faux "gap" type margin:
Screenshot 2021-10-26 at 11 28 21

Screenshot 2021-10-26 at 11 28 27

But the net result is that the combination of space-between, gap, and flex-basis: 33%; isn't really working to make the three items equally weighted. I could probably dive in and override some of the component CSS, but that also feels like the wrong approach. Any idea of a good fix?

**Comment by tellthemachines** [comment]
> I noticed in testing the styling doesn't seem quite right

Oh, that's weird, I can't reproduce it. It should just work, because it's using the Flex component. Though we might have to change that, because of:

> The only other piece I'd love for us to find a way to fix is making each item the same size

The problem with the Flex component is that FlexItem wraps an element around our buttons, so instead of having

```
<Flex>
    <button />
    <button />
</Flex>
```
which would allow us to style the buttons as flex children, we end up with

```
<Flex>
    <div>
        <button />
    </div>
    <div>
        <button />
    </div>
</Flex>
```
and because Flex is set up to use CSS-in-JS, we can't just hook some extra CSS to its classnames. We can pass the styles inline to FlexItem, but at that point it's easier (and much less messy) to not use Flex at all, and attach the CSS directly to our LinkControlTransforms markup.

Flex is still marked as experimental and if it is to be useful at all as a component it needs a fair bit of improvement. Not forcing a wrapper around the FlexItems would be a good start, as well as allowing customisation of individual item styles. To be honest I'm not sold on the idea of a component with the sole function of adding three lines of CSS to a div: it doesn't seem to solve any meaningful problem, and it makes the DX horrible (code bloat, wrappers around wrappers, having to hunt through multiple files to find where a style is declared). But I'm probably missing some context around this 🤷

Tl,dr: I think our best bet is to avoid using Flex for now, and add the styles manually.

File: packages/block-library/src/navigation-link/edit.js
Comment (tellthemachines) [reply packages/block-library/src/navigation-link/edit.js]: I'm seeing this as a micro-component that happens to be used in navigation link, but could potentially be used in any other block that has LinkControl as its main feature.

File: packages/block-library/src/navigation-link/edit.js
Comment (tellthemachines) [reply packages/block-library/src/navigation-link/edit.js]: Hmm, good point, what's the best way to check the blocks are registered?

**Comment by talldan** [comment]
> The problem with the Flex component is that FlexItem wraps an element around our buttons

Yeah, seems like FlexItem should have an `as` prop:
```jsx
<FlexItem as={ Button }>
```
But it doesn't right now. It always renders as a `View`, which is a `div`.

**Review by talldan — APPROVED** [review]
Working really well. I feel like it could do with a tiny bit more padding at the bottom underneath the icons, but it could just be me:
Screenshot 2021-10-27 at 4 24 19 pm

**Comment by tellthemachines** [comment]
> I feel like it could do with a tiny bit more padding at the bottom underneath the icons, but it could just be me

Hmm, adding margin or padding to the buttons or to their containers triggers a scrollbar on the popover 😕 
I'll merge this as is and we can work it out later as an enhancement.

**Comment by getdave** [comment]
Glad this worked out ❤️

**Pull request merged**
